Add more Info to locality load balancing page (#13682)
istio/istio.io: content/en/docs/tasks/traffic-management/locality-load-balancing/before-you-begin/index.md

test: yes
owner: istio/wg-networking-maintainers
---

> {{< tip >}}

Reviewer: one more nit 😄 I think it would be better to put this tip just before the `Environment Variables` section starts (i.e., after the setup diagram)

Author: I moved the tip before the `Environment Variables` section (after the setup diagram).

In a single multi-zone cluster environment, locality load balancing can also be configured for failover to a different zone within the same cluster.
To test it, you will need to create a cluster with multiple worker zones and deploy an istiod instance and the app to each zone.

1) If you don’t have a multi-zone Kubernetes cluster, you can deploy one locally using `kind` with the following command:

> {{< text syntax=bash snip_id=none >}}
> $ kind create cluster --config=- <<EOF
> kind: Cluster
> apiVersion: kind.x-k8s.io/v1alpha4
> nodes:
>   - role: control-plane
>   - role: worker
>   - role: worker
>   - role: worker
> EOF
> {{< /text >}}
> 
> 2) Use `topology.kubernetes.io/zone` to label each worker with a zone name:
> 
> {{< text syntax=bash snip_id=none >}}
> $ kubectl label node kind-worker topology.kubernetes.io/zone=us-south10
> $ kubectl label node kind-worker2 topology.kubernetes.io/zone=us-south12
> $ kubectl label node kind-worker3 topology.kubernetes.io/zone=us-south13
> {{< /text >}}

Reviewer: nit: a little better format for the commands:
```suggestion
    {{< text syntax=bash snip_id=none >}}
    $ kind create cluster --config=- <<EOF
    kind: Cluster
    apiVersion: kind.x-k8s.io/v1alpha4
    nodes:
    - role: control-plane
    - role: worker
    - role: worker
    - role: worker
    EOF
    {{< /text >}}

2) Use `topology.kubernetes.io/zone` to label each worker with a zone name:

    {{< text syntax=bash snip_id=none >}}
    $ kubectl label node kind-worker topology.kubernetes.io/zone=us-south10
    $ kubectl label node kind-worker2 topology.kubernetes.io/zone=us-south12
    $ kubectl label node kind-worker3 topology.kubernetes.io/zone=us-south13
    {{< /text >}}
```

Author: When I commit this change and run the page locally, I see some commands are not in the text block. I have tried with different indents but could not get the correct format. Any suggestions to this? 
![Screenshot 2023-09-05 at 11 09 25 AM](https://github.com/istio/istio.io/assets/36289669/85e0a8db-7964-4050-a392-f4802cacc95b)

> 

Reviewer: Would it be good to add a step 3 to tell users to deploy istiod and the app to the nodes? For example:
```suggestion

3) Deploy istiod to the control-plane node and the helloworld app to each of the worker nodes.

```

Author: Yes, sounds good. Should I provide any commands for step 3?

Reviewer: I don't think it's really necessary. I probably wouldn't do it unless there was a very simple one line thing you could show to help. Otherwise it's too much detail for a tip, and users should be able to figure it out for themselves IMO.

Author: Okay, agreed.

    caption="Setup for locality load balancing tasks"
    >}}

{{< tip >}}
In a single multi-zone cluster environment, locality load balancing can also be configured for failover to a different zone within the same cluster.
To test it, you will need to create a cluster with multiple worker zones and deploy an istiod instance and the app to each zone.

> 1) If you don’t have a multi-zone Kubernetes cluster, you can deploy one locally using `kind` with the following command:
> 
>   {{< text syntax=bash snip_id=none >}}
>     $ kind create cluster --config=- <<EOF
>     kind: Cluster
>     apiVersion: kind.x-k8s.io/v1alpha4
>     nodes:
>     - role: control-plane
>     - role: worker
>     - role: worker
>     - role: worker
>     EOF
>   {{< /text >}}
> 
> 2) Use `topology.kubernetes.io/zone` to label each worker with a zone name:
> 
>   {{< text syntax=bash snip_id=none >}}
>     $ kubectl label node kind-worker topology.kubernetes.io/zone=us-south10
>     $ kubectl label node kind-worker2 topology.kubernetes.io/zone=us-south12
>     $ kubectl label node kind-worker3 topology.kubernetes.io/zone=us-south13
>   {{< /text >}}
> 
> 3) Deploy istiod to the control-plane node and the helloworld app to each of the worker nodes.

Reviewer: Looks like there is a limitation or bug in the page generator for list entries with indented commands inside a tip 😞

Here's my best suggestion given the limitation:

```suggestion
1: If you don’t have a multi-zone Kubernetes cluster, you can deploy one locally using `kind` with the following command:

{{< text syntax=bash snip_id=none >}}
$ kind create cluster --config=- <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
- role: worker
- role: worker
EOF
{{< /text >}}

2: Use `topology.kubernetes.io/zone` to label each worker with a zone name:

{{< text syntax=bash snip_id=none >}}
$ kubectl label node kind-worker topology.kubernetes.io/zone=us-south10
$ kubectl label node kind-worker2 topology.kubernetes.io/zone=us-south12
$ kubectl label node kind-worker3 topology.kubernetes.io/zone=us-south13
{{< /text >}}

3: Deploy istiod to the control-plane node and the helloworld app to each of the worker nodes.
```

Author: Thank you! Below is what it looks like after this commit.
![Screenshot 2023-09-05 at 12 09 58 PM](https://github.com/istio/istio.io/assets/36289669/be6f9ada-a539-43db-af65-f020744d5842)

mutually exclusive. Attempting to configure both may lead to unexpected
behavior.
{{< /warning >}}

> {{< tip >}}

Author: I moved the tip to the start of the `before you begin` section.

In a single multi-zone cluster environment, locality load balancing can also be configured for failover to a different zone within the same cluster.
To test it, you will need to create a cluster with multiple worker zones and deploy an istiod instance and the app to each zone.

1) If you don’t have a multi-zone Kubernetes cluster, you can deploy one locally using `kind` with the following command:

> {{< text bash >}}

Reviewer: Ideally we would test these snips, but since the test framework doesn't really support customizing the test cluster like this and since it's just a tip, you should probably suppress generation of the snips.
```suggestion
{{< text syntax=bash snip_id=none >}}
```

Author: updated

$ kind create cluster --config=- <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
  - role: worker
  - role: worker
  - role: worker
EOF
{{< /text >}}

2) Use `topology.kubernetes.io/zone` to label each worker with a zone name:

> {{< text bash >}}

Reviewer: ```suggestion
{{< text syntax=bash snip_id=none >}}
```

mutually exclusive. Attempting to configure both may lead to unexpected
behavior.
{{< /warning >}}

{{< tip >}}

> In a single multi-zone cluster environment, locality load balancing can also be configured for failover to a different zone within the same cluster. To test it, you will need to create a cluster with multiple worker zones and deploy istiod instance and the app to each zone. The following example shows how to set up a kind cluster with multiple worker zones. The remaining steps are identical to this multicluster example.
> 
> Generate yaml to define a cluster with 1 control-plane and 3 worker nodes:
> 
> {{< text bash >}}
> $ cat <<EOF > kind-cluster-config.yaml
> kind: Cluster
> apiVersion: kind.x-k8s.io/v1alpha4
> nodes:
>   - role: control-plane
>   - role: worker
>   - role: worker
>   - role: worker
> EOF
> {{< /text >}}
> 
> Create a kind cluster with yaml generated from above:
> 
> {{< text bash >}}
> $ kind create cluster --config kind-cluster-config.yaml
> {{< /text >}}
> 
> Use `topology.kubernetes.io/zone` to label each worker with a zone name:

Reviewer: Maybe make the tip a little shorter?

```suggestion
In a single multi-zone cluster environment, locality load balancing can also be configured for failover to a different zone within the same cluster.
To test it, you will need to create a cluster with multiple worker zones and deploy an istiod instance and the app to each zone.

1) If you don’t have a multi-zone Kubernetes cluster, you can deploy one locally using `kind` with the following command:

{{< text bash >}}
$ kind create cluster --config=- <<EOF
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
  - role: control-plane
  - role: worker
  - role: worker
  - role: worker
EOF
{{< /text >}}

2) Use `topology.kubernetes.io/zone` to label each worker with a zone name:
```

Author: Thank you, @frankbu. I committed your suggestion.
